### PR TITLE
Consolidating keystrokes

### PR DESCRIPTION
**Pull request by bartcull**
@nartiles, here's my first pass at consolidating keystrokes to limit the number of searches being triggered by the UISearchController.

### CODE REVIEW

File: WikiDrips/WikiTableViewController.swift
Comment (nartiles) [inline WikiDrips/WikiTableViewController.swift]: If we have coalescing I think we should remove the character limit. I also find it (as a user) limiting that I can't search for short words (i.e. like Pi)

**Review by nartiles — CHANGES_REQUESTED** [review]
Looks good, I would also set the `.tolerance` property on `Timer`, it helps with power requirements.

File: WikiDrips/WikiTableViewController.swift
Comment (nartiles) [inline WikiDrips/WikiTableViewController.swift]: If it's empty, wouldn't you want to cancel any running operations also?

Comment (bartcull) [inline WikiDrips/WikiTableViewController.swift]: Good catch.

**Comment by bartcull** [comment]
@nartiles, good catch.

**Review by nartiles — APPROVED** [review]

**Comment by bartcull** [comment]
@nartiles, I had to change things a bit because of a merge conflict.  Still look good?

File: WikiDrips/WikiTableViewController.swift
Comment (nartiles) [inline WikiDrips/WikiTableViewController.swift]: I hadn't seen this before. I think you do need `cancelPrefetchingForRowsAt` when you're scrolling a long list quickly then I think you'll get this method called when on rows that the table view no longer needs.

**Review by nartiles — CHANGES_REQUESTED** [review]

File: WikiDrips/WikiTableViewController.swift
Comment (bartcull) [inline WikiDrips/WikiTableViewController.swift]: So sounds to me like something I should handle in a pr that adds the ability to retrieve more than 10 records.  A prefetch for the last record could call the Wikipedia api for the next set of 10 records.  Sound right?

**Comment by bartcull** [comment]
@nartiles, I'm curious if the git log history for this branch looks okay to you.  I prefer to rebase to keep things a bit cleaner, but (unless I'm mistaken) that would risk losing your comments.  Anyway, just want to make sure I'm handling multiple branches the way you'd like.

File: WikiDrips/WikiTableViewController.swift
Comment (nartiles) [inline WikiDrips/WikiTableViewController.swift]: A few things.. first, I think fetching only the first 10 records is too low. I think we the default page size should fill at least 1 screen full.. so I don't know if you can increase the page size, but if you can I would change it to 25 or 50. From a 'device power' perspective its better to fetch more data and make less requests.
Second, regarding pagination, I'd be curious to see how the scrolling feels with your approach (i.e. using the prefetch of the last record to trigger fetching the next page). On LSM Eldar didn't use prefetch but instead used the cellForRowAtIndex method. However, he didn't wait till we got to the last row, instead he triggers the next page fetch when we get a certain fraction down the first page.
I'm ok if you want to implement the cancelPrefetching in that PR instead.

**Comment by nartiles** [comment]
Regarding branches, yes, we usually rebase to keep things cleaner but what you did hear is fine too. My big no-no is having dangling branches or duplicate commits on different branches, then it gets harder to figure out when the feature actually made it into master. In our flow, after code review is approved then the ticket goes to this 'Waiting' column on the board which is an opportunity for the dev of that ticket to rebase if necessary before going to 'Waiting for QA'.

On a related note though, two observations on your commit messages:
1. I've heard, and I tend to agree, that commit messages should be in the present tense instead of past tense. The idea being that the log should sound like a sequence of commands that you're issuing. Not a big deal in my opinion, just something Eldar and I have settled on with LSM.
2. I think I'd prefer if we're always explicit about what changes went into a commit, specially after PR changes. So instead of 'fixes for 5 pr comments', try to be explicit and group them into meaningful change commits. This is an area where I found a git GUI (such as Source Tree) hugely valuable, (and I have no idea how to do this from command line) because the GUI lets me stage/unstage individual lines which you sometimes need to do to make the set of commits be logically grouped.

**Comment by bartcull** [comment]
Cool.  Those guidelines for commits sound good to me.  

I'll work on prefetching in a separate pr.  So I'll merge this pr sometime today unless you think something is still missing.

**Pull request merged**
